File: backend/app/schemas/statistics.py
"""Phase 5.1: Statistics response schemas."""
from pydantic import BaseModel


# ── Score distribution bucket ────────────────────────────────────────────────

class ScoreDistributionBucket(BaseModel):
    range: str  # e.g. "<30", "30-39", "40-59", "60-69", "70-84", "85-100", "100-150"
    count: int


# ── Class-course statistics ──────────────────────────────────────────────────

class ClassCourseStatistics(BaseModel):
    class_id: int
    class_name: str
    course_id: int
    course_name: str
    semester_id: int
    student_count: int
    avg_score: float
    max_score: float
    min_score: float
    ultra_low_rate: float   # < 30
    low_rate: float         # < 40
    pass_rate: float        # >= 60
    good_rate: float        # >= 70
    excellent_rate: float   # >= 85
    distribution: list[ScoreDistributionBucket]


# ── Class-course ranking (single subject) ────────────────────────────────────

class ClassRankingItem(BaseModel):
    rank: int
    student_id: int
    student_no: str
    student_name: str
    score: float

class ClassRankingResponse(BaseModel):
    items: list[ClassRankingItem]
    total: int
    page: int
    page_size: int


# ── Grade ranking (total score across subjects) ─────────────────────────────

class GradeRankingSubjectScore(BaseModel):
    course_id: int
    course_name: str
    score: float

class GradeRankingItem(BaseModel):
    rank: int
    student_id: int
    student_no: str
    student_name: str
    class_name: str
    total_score: float
    subjects: list[GradeRankingSubjectScore]

class GradeRankingResponse(BaseModel):
    items: list[GradeRankingItem]
    total: int
    page: int
    page_size: int


# ── Student personal statistics ──────────────────────────────────────────────

class StudentSubjectStat(BaseModel):
    course_id: int
    course_name: str
    score: float
    class_rank: int
    class_total: int
    grade_rank: int
    grade_total: int
    class_avg: float

class StudentStatisticsResponse(BaseModel):
    student_id: int
    student_no: str
    student_name: str
    class_name: str
    semester_id: int
    subjects: list[StudentSubjectStat]
    total_score: float
    total_class_rank: int
    total_grade_rank: int


# ── Score trend (multi-semester) ─────────────────────────────────────────────

class TrendPoint(BaseModel):
    semester_id: int
    semester_name: str
    student_score: float | None
    class_avg: float | None

class ScoreTrendResponse(BaseModel):
    course_id: int
    course_name: str
    trend: list[TrendPoint]


# ── Cross-class comparison within a grade ────────────────────────────────────

class ClassComparisonItem(BaseModel):
    class_id: int
    class_name: str
    avg_score: float
    pass_rate: float
    excellent_rate: float
    student_count: int

class GradeComparisonResponse(BaseModel):
    course_id: int
    course_name: str
    semester_id: int
    classes: list[ClassComparisonItem]

File: frontend/src/api/statistics.ts
import http from './http'

// ── Types ───────────────────────────────────────────────────────────────────

export interface ScoreDistributionBucket {
  range: string
  count: number
}

export interface ClassCourseStatistics {
  class_id: number
  class_name: string
  course_id: number
  course_name: string
  semester_id: number
  student_count: number
  avg_score: number
  max_score: number
  min_score: number
  ultra_low_rate: number
  low_rate: number
  pass_rate: number
  good_rate: number
  excellent_rate: number
  distribution: ScoreDistributionBucket[]
}

export interface ClassRankingItem {
  rank: number
  student_id: number
  student_no: string
  student_name: string
  score: number
}

export interface ClassRankingResponse {
  items: ClassRankingItem[]
  total: number
  page: number
  page_size: number
}

export interface GradeRankingSubjectScore {
  course_id: number
  course_name: string
  score: number
}

export interface GradeRankingItem {
  rank: number
  student_id: number
  student_no: string
  student_name: string
  class_name: string
  total_score: number
  subjects: GradeRankingSubjectScore[]
}

export interface GradeRankingResponse {
  items: GradeRankingItem[]
  total: number
  page: number
  page_size: number
}

export interface StudentSubjectStat {
  course_id: number
  course_name: string
  score: number
  class_rank: number
  class_total: number
  grade_rank: number
  grade_total: number
  class_avg: number
}

export interface StudentStatisticsResponse {
  student_id: number
  student_no: string
  student_name: string
  class_name: string
  semester_id: number
  subjects: StudentSubjectStat[]
  total_score: number
  total_class_rank: number
  total_grade_rank: number
}

export interface TrendPoint {
  semester_id: number
  semester_name: string
  student_score: number | null
  class_avg: number | null
}

export interface ScoreTrendResponse {
  course_id: number
  course_name: string
  trend: TrendPoint[]
}

export interface ClassComparisonItem {
  class_id: number
  class_name: string
  avg_score: number
  pass_rate: number
  excellent_rate: number
  student_count: number
}

export interface GradeComparisonResponse {
  course_id: number
  course_name: string
  semester_id: number
  classes: ClassComparisonItem[]
}

// ── API calls ───────────────────────────────────────────────────────────────

export function fetchClassCourseStats(params: {
  class_id: number
  course_id: number
  semester_id: number
}) {
  return http.get<ClassCourseStatistics>('/statistics/class-course', { params })
}

export function fetchClassRanking(params: {
  class_id: number
  course_id: number
  semester_id: number
  page?: number
  page_size?: number
}) {
  return http.get<ClassRankingResponse>('/statistics/class-ranking', { params })
}

export function fetchGradeRanking(params: {
  grade_id: number
  semester_id: number
  page?: number
  page_size?: number
}) {
  return http.get<GradeRankingResponse>('/statistics/grade-ranking', { params })
}

export function fetchStudentStatistics(
  studentId: number,
  params: { semester_id: number },
) {
  return http.get<StudentStatisticsResponse>(`/statistics/student/${studentId}`, { params })
}

export function fetchScoreTrend(params: {
  student_id: number
  course_id: number
}) {
  return http.get<ScoreTrendResponse>('/statistics/trend', { params })
}

export function fetchGradeComparison(params: {
  grade_id: number
  course_id: number
  semester_id: number
}) {
  return http.get<GradeComparisonResponse>('/statistics/comparison', { params })
}
